--- a/nptl/allocatestack.c
+++ b/nptl/allocatestack.c
@@ -210,7 +210,7 @@ advise_stack_range (void *mem, size_t size, uintptr_t pd, size_t guardsize)
    new stack or reusing a cached stack of sufficient size.
    ATTR must be non-NULL and point to a valid pthread_attr.
    PDP must be non-NULL.  */
-static int
+int
 allocate_stack (const struct pthread_attr *attr, struct pthread **pdp,
                void **stack, size_t *stacksize)
 {
